apps/prairielearn/src/ee/pages/studentCourseInstanceUpgrade/studentCourseInstanceUpgrade.ts

```diff
@@ -17,7 +44,222 @@ router.get(
       return;
     }
 
-    res.send(StudentCourseInstanceUpgrade({ resLocals: res.locals }));
+    const institution = InstitutionSchema.parse(res.locals.institution);
+    const course = CourseSchema.parse(res.locals.course);
+    const course_instance = CourseInstanceSchema.parse(res.locals.course_instance);
+    const user = UserSchema.parse(res.locals.authn_user);
+
+    const planGrants = await getPlanGrantsForPartialContexts({
+      institution_id: institution.id,
+      course_instance_id: course_instance.id,
+      user_id: user.user_id,
+    });
+    const requiredPlans = await getRequiredPlansForCourseInstance(res.locals.course_instance.id);
+    const missingPlans = getMissingPlanGrants(planGrants, requiredPlans);
+
+    // TODO: fetch pricing information from Stripe API; cache it too.
+
+    res.send(
+      StudentCourseInstanceUpgrade({
+        course,
+        course_instance,
+        missingPlans,
+        resLocals: res.locals,
+      }),
+    );
+  }),
+);
+
+const UpgradeBodySchema = z.object({
+  terms_agreement: z.literal('1').optional(),
+  unsafe_plan_names: z.string(),
+});
+
+// Only a subset of all plans are allowed to be paid for on this page.
+const PlanNamesSchema = z.array(z.enum(['basic', 'compute']));
+
+router.post(
+  '/',
+  asyncHandler(async (req, res) => {
+    if (req.body.__action === 'upgrade') {
+      console.log(res.locals);
+      const institution = InstitutionSchema.parse(res.locals.institution);
+      const course = CourseSchema.parse(res.locals.course);
+      const course_instance = CourseInstanceSchema.parse(res.locals.course_instance);
+      const user = UserSchema.parse(res.locals.authn_user);
+
+      const body = UpgradeBodySchema.parse(req.body);
+
+      if (!body.terms_agreement) {
+        throw error.make(400, 'You must agree to the terms and conditions.');
+      }
+
+      const rawPlanNames = body.unsafe_plan_names.split(',');
+      const planNames = PlanNamesSchema.parse(rawPlanNames);
+
+      // TODO: should we use lookup keys instead? See
+      // https://stripe.com/docs/products-prices/manage-prices#lookup-keys
+      // Unfortunately, these can't be set or modified from the Stripe console,
+      // so they aren't as useful as they could be.
```

Comment:
I'm chatting with Stripe support about this. It wouldn't be the worst thing to have price IDs hardcoded into our config; it just means we'd need to redeploy if we ever need to update prices.

Reply:
Another option seems to be to hardcode _product_ IDs in our config, and then look up the default prices for them at runtime.

Reply:
I'm not quite sure I follow how this will work with the current code. We need to separately create products/prices in stripe (which will give us IDs?) and then put these IDs in our config?

Reply:
Yeah, so, some background: Stripe Checkout requires line items to reference a [price](https://stripe.com/docs/api/prices), and prices in turn must be associated with products. The way things are currently implemented, we have to manually set price IDs for each plan in `config.json`. The problem with this is that once a price has been used for a purchase, it becomes immutable, so changing the price would require creating a new one, updating the config, and redeploying. This is not the worst thing, as prices shouldn't generally change to much...

Reply:
Discussed with Matt: we'll use Product IDs in our config and look up the default price from those. If we want to get fancy, we can listen for webhooks telling us when the default price changed and clear the cached price.

Reply:
Did this get implemented?

Reply:
Just pushed it: c310daa.